Create /map/ route for wiki pages

nstjean opened the pull request:
Fixes #7060  (<=== Add issue number here)

- Created a route /map/wiki/:wikislug - I couldn't use /map/:wikislug because we already had a route with that pattern.

- If Wiki exists then it is checking for location info to display:

![🎈 Public Lab: Wiki - Google Chrome_007](https://user-images.githubusercontent.com/49460529/71598498-0962d100-2b15-11ea-92c0-f7feaec7f5ac.png)


- If the map is panned while looking at a wiki page map then it displays the current location in the hash so it can be copied and pasted. If there is location info in the URL hash it will override the display of the wiki's saved location.

![🎈 Public Lab: Wiki - Google Chrome_010](https://user-images.githubusercontent.com/49460529/71598596-75453980-2b15-11ea-9aa0-8e2f02315fec.png)


- If Wiki exists but does not have location information this is displayed:

![🎈 Public Lab: Wiki - Google Chrome_008](https://user-images.githubusercontent.com/49460529/71598504-1253a280-2b15-11ea-82d8-ccc3b71e346e.png)


- If wiki does not exist user is redirected to the standard /map/ and a "not found" message is shown:

![🎈 Public Lab: Map - Google Chrome_009](https://user-images.githubusercontent.com/49460529/71598509-17b0ed00-2b15-11ea-80a6-f2f32eab45df.png)

nstjean commented:
Here is a wiki page that has a location:

![🎈 Public Lab: Events - Google Chrome_011](https://user-images.githubusercontent.com/49460529/71631508-08927380-2bd7-11ea-90d8-fdc5af8ff0f2.png)


This is the map when the button is clicked on:

![🎈 Public Lab: Wiki - Google Chrome_017](https://user-images.githubusercontent.com/49460529/71631516-1516cc00-2bd7-11ea-8501-00d09549830e.png)


This is a note that has a location with a top map:

![🎈 Public Lab: This is a post testing location - Google Chrome_012](https://user-images.githubusercontent.com/49460529/71631537-27910580-2bd7-11ea-80f0-10bfc362b444.png)


This is the map when the button has been clicked on:

![🎈 Public Lab: Map - Google Chrome_013](https://user-images.githubusercontent.com/49460529/71631552-3a0b3f00-2bd7-11ea-9701-cc42a52ca661.png)


And this is how the url updates while you are panning and zooming!

![Peek 2019-12-31 14-11](https://user-images.githubusercontent.com/49460529/71631614-8ce4f680-2bd7-11ea-8aad-500cfdb2b5fa.gif)

nstjean commented:
Based on feedback in https://github.com/publiclab/plots2/issues/6993 I've changed the routes and URL hash pattern
Wiki route is now `/map/:wikislug`
And general route is `/map#zoom/lat/lon`

Currently there is no way to specify active layers, that can be added on to the hash if it is needed!

sagarpreet-chadha commented:
Looks great!!!

sagarpreet-chadha commented:
- [ ] Make a “/map/_wiki_slug_” route that sources lat/lon/zoom from the matching wiki page
- [ ] Generate/update URL hashes on /map/_wiki_slug_#zoom/lat/lon pages as you drag the
map
- [ ] Display notes, wikis, and people on “/map/_slug_” pages
- [ ] Display popups showing title, link for each wiki, note, and question
- [ ] Make a “#” permalink (in grey) appear in wiki page popups that show “/map/_slug_”
permalinks for sharing
- [ ] Add a “View map” button to geotagged page sidebars, opening “/map/_slug_” style URLs
for wikis, and “/map#zoom/lat/lon” URL hashes for notes
- [ ] Add a “Add to map +” prompt button on any “/map/...” pages to add a note, question or
wiki at that location
- [ ] Direct users to a post type selection page from the above prompt button, at
“/post/choose?tags=lat:10,lon:10 -- preserving params!)

sagarpreet-chadha commented:
Hey @nstjean , i think we will be solving all these things here, right?

nstjean commented:
@sagarpreet-chadha I've marked off the 3 that are taken care of with my current PRs!

nstjean commented:
@jywarren I created this branch from the #7046 branch, so I think that needs to be merged first, then this rebased to master.

nstjean commented:
@jywarren This is all set now!

jywarren commented:
Ah, i'm sorry, i wanted to add one more small request - could we left-align the button? Apologies! Then this is perfect!

nstjean commented:
Yes of course! I'll get that moved.

nstjean commented:
Here you go, it's all set!

![FireShot Capture 165 - 🎈 Public Lab_ Software Outreach - localhost](https://user-images.githubusercontent.com/49460529/71732521-76c47980-2e15-11ea-9d39-ba00b90c4cc7.png)

nstjean commented:
Now the codecov test is failing. Do I need to do something different?

jywarren commented:
Hmm, I don't think so - i want to see if we can modify the thresholds; i think it's saying that it doesn't like that the new code is not tested to at least the percentage coverage of the overall repository. We could add a functional test or a system test to address this, maybe? We're just getting used to the new CodeCov checks, sorry!

nstjean commented:
Hmmm ok. I was going to add some tests for this anyways!

nstjean commented:
Ok I added map_test.rb to system tests to test out various map options and display. It should cover the map#wiki controller.

sagarpreet-chadha commented on app/views/map/map.html.erb:
```diff
@@ -21,53 +20,63 @@
         new L.LatLng(-54.36775852, 178.59375)
       );
 
-      var user_lat = null;
```
Yes awesome i wanted this to be removed from here and not the URLHash logic! Thanks!

sagarpreet-chadha commented on app/assets/javascripts/urlMapHash.js:
```diff
@@ -0,0 +1,53 @@
+function urlMapHash() {
+  // This is based off of jywarren's urlhash, made specific to our map hash needs
+
```
Hey this is cool, can you tell what specific changes you did to match plots2? Is the URL different now? Thanks!

nstjean replied:
I changed the format to be `#zoom/lat/lon`.  I really like having a hash function available to get and set the values!

crisner replied:
Quick question, have the layer names been removed from the hash?

nstjean replied:
Currently yes they aren't in the hash. It had been set up to use the route for the layer names, which had to be changed. Do you need them added back in? It's easy enough to add them at the end of the hash, if that works for @jywarren ?

crisner replied:
No, that's fine for now. I was trying to figure out an issue when I noticed we don't have them anymore. :)

crisner replied:
We had them to track active layers. In LEL it is used in the embed to preserve the state of the layers. I don't remember if this was the case in plots2? @sagarpreet-chadha, could you confirm?

sagarpreet-chadha replied:
Hey,
So i made embed feature to take array of layers as input, and show them on map by default (@nstjean ...we will use this in /people map in plots2). This feature is unrelated to url hash AFAIK.

If we will require layer names in future, it should be pretty easy, right?  Thanks!

crisner replied:
Okay. Got it. To correct what I said earlier, it wasn't in the hash but before it. Thanks, @sagarpreet-chadha!

nstjean commented:
Wow, I don't know what was changed with the codecov, but now my patch went from 70% before to 15% now!

jywarren commented:
Ah, i think it gets logged gradually, so by the time the Travis run finishes it may be much higher. I wish it notified about that! Let's see!

nstjean commented:
It did go back up! Interesting.

This is the test failure, which doesn't make sense to me. Unless me adding the new tags broke this somehow? I didn't change anything that existed, just added.
![FireShot Capture 168 - Job #15870 3 - publiclab_plots2 - Travis CI - travis-ci org](https://user-images.githubusercontent.com/49460529/71828356-31e84f00-3070-11ea-913e-1e637f573d92.png)

And then every one of my map tests is timing out and erroring. :unamused:

nstjean commented:
I took a look and ran it in my local dev to try out some different numbers.

I believe that page may be broken. Take a look: https://publiclab.org/contributors/general
How can we have 61 contributors making 32 posts? I think me adding some new tags in the test fixtures caused this to be visible. The test is supposed to count the number of users in the controller vs. the number of rows on the page, and it's not equal: it's coming back with 0 rows but 1 user in the controller.
https://github.com/publiclab/plots2/blob/df0b944da3164602558cf2230723671454415dbf/test/functional/tag_controller_test.rb#L372-L385

What do you want me to do? Fixing this page could take a while.

nstjean commented:
Oh scratch that! It's the fact that I added a power tag. `zoom:13` causes a failure, as does `lat:41.87`. But `balloon` `blog` and `awesome` all pass. So I've changed the test to check for `blog` instead of Tag.last.name so it won't run into this.

nstjean commented:
So it looks like I fixed that TagController error, but all the map tests are timing out still. They run fine on my local env.

jywarren commented:
Hm. Can you look at the screenshot tests that snap pics of pages with maps?
Maybe we adjusted the timing significantly on those to make them work.

On Mon, Jan 6, 2020 at 12:51 PM Natalie St Jean <notifications@github.com>
wrote:

> So it looks like I fixed that TagController error, but all the map tests
> are timing out still. They run fine on my local env.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/plots2/pull/7061?email_source=notifications&email_token=AAAF6JYYZFVLLGIRVGUN5MLQ4NVQVA5CNFSM4KBPX5XKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEIGGWNQ#issuecomment-571239222>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAF6J473QS5TWUP4IEON7DQ4NVQVANCNFSM4KBPX5XA>
> .
>

nstjean commented:
> Hm. Can you look at the screenshot tests that snap pics of pages with maps? Maybe we adjusted the timing significantly on those to make them work.
> […](#)
> On Mon, Jan 6, 2020 at 12:51 PM Natalie St Jean ***@***.***> wrote: So it looks like I fixed that TagController error, but all the map tests are timing out still. They run fine on my local env. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <#7061?email_source=notifications&email_token=AAAF6JYYZFVLLGIRVGUN5MLQ4NVQVA5CNFSM4KBPX5XKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEIGGWNQ#issuecomment-571239222>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAAF6J473QS5TWUP4IEON7DQ4NVQVANCNFSM4KBPX5XA> .

I only see one that involves a map and it doesn't have any different time limit.

I'm going to try re-writing some tests as controller tests and leaving only the necessary as system tests. Maybe that will help.

nstjean commented:
Wow, it won't even take a screenshot of the map page.
![FireShot Capture 176 - Job #15911 5 - publiclab_plots2 - Travis CI - travis-ci org](https://user-images.githubusercontent.com/49460529/71905927-1bf19180-3137-11ea-9267-baa155f29fb0.png)

nstjean commented:
I removed the screenshot of the map page just to see what would happen with the rest. 

![FireShot Capture 180 - Job #15916 5 - publiclab_plots2 - Travis CI - travis-ci org](https://user-images.githubusercontent.com/49460529/71909512-2fecc180-313e-11ea-9490-1afee50ed85d.png)

Weird.

sagarpreet-chadha commented:
The tests are passing correctly on your local, right?

> session deleted because of tab crash 

I think there is a memory isssue when running headless chrome.

sagarpreet-chadha commented:
I am closing and re-opening this PR to see if same tests are failing or not.

nstjean commented:
> The tests are passing correctly on your local, right?

Yes they all work fine on local. I know the maps are pretty big and slow.

nstjean commented:
@SidharthBansal This says it's still running??

SidharthBansal commented:
I think Jeff or Sagar can help you here. I may misguide you. I am unsure here.

cesswairimu commented:
I see its a timeout error...Seems like travis is taking too long to run, checking if this is affecting any other PRs

cesswairimu commented:
Seems like its the only pull request getting this error...saw this `Builds can unfortunately time out, either during installation of dependencies, or during the build itself, for instance because of a command that’s taking a longer amount of time to run while not producing any output` actually last build took 50min :thinking:  @nstjean do you know after which commit the build started failing we try to debug from there?

nstjean commented:
Yes it was when I added the map tests in https://github.com/publiclab/plots2/pull/7061/files#diff-5cf134f356cb4ba3ed15181275b5032b - that's where the test build is crashing too. But I don't know why, it runs fine on my local dev setup. I can comment those tests out to confirm.

cesswairimu commented on test/system/map_test.rb:
```diff
@@ -3,35 +3,33 @@
-    assert_equal(-25, page.evaluate_script("Math.round(map.getCenter().lat)"))
-    assert_equal(-13, page.evaluate_script("Math.round(map.getCenter().lng)"))
-    assert_equal(9, page.evaluate_script("map.getZoom()"))
+  #   assert_equal(-25, page.evaluate_script("Math.round(map.getCenter().lat)"))
```
I have a long shot thought that maybe the calculation inside the assertion could be failing this...
Do you mind trying doing something like this for one of the assertion to see if it passes?  
```
lat = page.evaluate_script("Math.round(map.getCenter().lat)")
assert_equal(-25, lat)?
```

nstjean commented:
Well it passes with those tests commented out. I'll try your suggestion @cesswairimu !

cesswairimu commented:
Yeah, you are right, I am taking a look at the tests to see if there is any optimization that can be done to help them run faster

cesswairimu commented:
Well, that did not work..its actually failing on this line `visit '/map#9/-25/-13'`

nstjean commented:
Ok. Hiding that test and running a different test to see if it's just that one line.

nstjean commented:
Guess not. It doesn't like any part of this map in the test.

cesswairimu commented:
I will make a branch off of this and play with the test see if I can get the build to pass

cesswairimu commented on test/system/map_test.rb:
```diff
@@ -0,0 +1,37 @@
+# https://guides.rubyonrails.org/testing.html#implementing-a-system-test
+
+class MapTest < ApplicationSystemTestCase
+  Capybara.default_max_wait_time = 60
```
Hey @nstjean, increasing the wait time seems to do the trick.. I changed this to` Capybara.default_max_wait_time = 120`

cesswairimu replied:
actually 80 works too

nstjean replied:
Ok I'm trying with 120 just to be safe. ;) If that works I'll change it to 80.

nstjean commented:
Hooray!! @jywarren this passed the tests! I think it's ready for merge!

jywarren commented:
Awesome!! we can test it at stable.publiclab.org now.

nstjean commented:
Yay!!! Oh this is exciting to see, haha.
